Add advanced DICOM client connection/assocation

amoerie opened the pull request:
Fixes #1096
Fixes #985
Fixes #1090 

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] I have updated API documentation
- [x] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- [x] Add a new advanced DICOM client which exposes controls to manipulate the underlying assocation
- [x] Rewrite the existing DICOM client on top of the advanced DICOM client
  - [x] Open associations and connections using the advanced DICOM client
  - [x] Implement all the special DicomClient behavior: lingering, max queries/assoc, etc.
  - [x] Remove the old state-based implementation
- [x] Add XML documentation to all new public members
- [x] Add copyright to all new files
- [x] Check for breaking public API changes on `DicomClient`
- [x] Add markdown documentation on how to use the new advanced DICOM client
- [x] Add integration tests for some common paths for the new advanced DICOM client
- [x] Add unit tests for some of the underlying infrastructure of the advanced DICOM client

I am creating this pull request already to discuss the design. 

The signature currently looks like this (updated February 2022):

```csharp  
    public interface IAdvancedDicomClientConnectionFactory
    {
       Task<IAdvancedDicomClientConnection> OpenConnectionAsync(AdvancedDicomClientConnectionRequest request, CancellationToken cancellationToken);
    }

    public interface IAdvancedDicomClientConnection : IDicomClientConnection
    {
        Task<IAdvancedDicomClientAssociation> OpenAssociationAsync(AdvancedDicomClientAssociationRequest request, CancellationToken cancellationToken);
    }

    public interface IAdvancedDicomClientAssociation: IDisposable
    {
        DicomAssociation Association { get; }

        IAsyncEnumerable<DicomResponse> SendRequestAsync(DicomRequest dicomRequest, CancellationToken cancellationToken);
        ValueTask ReleaseAsync(CancellationToken cancellationToken);
        ValueTask AbortAsync(CancellationToken cancellationToken);
    }
```

The usage currently looks like this (updated February 2022)

```csharp
var cancellationToken = CancellationToken.None;
// Alternatively, inject IDicomServerFactory via dependency injection instead of using this static method
using var server = DicomServerFactory.Create<DicomCEchoProvider>(12345); 

var connectionRequest = new AdvancedDicomClientConnectionRequest
{
    NetworkStreamCreationOptions = new NetworkStreamCreationOptions
    {
        Host = "127.0.0.1",
        Port = server.Port,
    }
};

// Alternatively, inject IAdvancedDicomClientConnectionFactory via dependency injection instead of using this static method
using var connection = await AdvancedDicomClientConnectionFactory.OpenConnectionAsync(connectionRequest, cancellationToken);

var associationRequest = new AdvancedDicomClientAssociationRequest
{
    CallingAE = "EchoSCU",
    CalledAE = "EchoSCP"
};

var cEchoRequest = new DicomCEchoRequest();

using var association = await connection.OpenAssociationAsync(associationRequest, cancellationToken);
try
{
    DicomCEchoResponse cEchoResponse = await association.SendCEchoRequestAsync(cEchoRequest, cancellationToken).ConfigureAwait(false);
    
    Console.WriteLine(cEchoResponse.Status);
}
finally
{
    await association.ReleaseAsync(cancellationToken);
}
```

@gofal what do you think? If you like this, we can make a few extra steps:

1. Implement full suite of tests, that already cover the old DicomClient
2. Implement proper logging in AdvancedDicomClient
3. Reimplement DicomClient on top of AdvancedDicomClient

I think this advanced DICOM client 

- would solve the stack trace issues (long stack traces due to state machine implementation)
- would solve the use case for power users that wish to send thousands of requests (they can now manage their associations how they wish)
- would solve the use case for power users that wish to implement advanced scenarios (an interceptor is provided that can inject custom behavior before each request or response)
- ... to be continued

amoerie committed:
Add global json that enforces using .NET Core 3.1 SDK

amoerie committed:
Only use .NET Core 3.1 in FO-DICOM.AspNetCore

amoerie committed:
Add initial prototype of new advanced DICOM client

amoerie commented:
This little unit test shows how the new advanced DICOM client can be used:

       ```csharp
 [Fact]
        public async Task SendAsync_SingleRequest_Recognized()
        {
            var server = "127.0.0.1";
            var port = Ports.GetNext();
            var callingAE = "SCU";
            var calledAE = "ANY-SCP";

            using (CreateServer<DicomCEchoProvider>(port))
            {
                var client = CreateClient();

                // TODO Alex perhaps a fluent API here?
                var openAssociationRequest = new OpenAssociationRequest
                {
                    ConnectionToOpen = new OpenConnectionRequest
                    {
                        NetworkStreamCreationOptions = new NetworkStreamCreationOptions
                        {
                            Host = server,
                            Port = port,
                        },
                        Logger = _logger.IncludePrefix(nameof(IAdvancedDicomClient)),
                        FallbackEncoding = DicomEncoding.Default,
                        DicomServiceOptions = new DicomServiceOptions()
                    },
                    AssociationToOpen = new DicomAssociation(callingAE, calledAE)
                    {
                        RemoteHost = server,
                        RemotePort = port,
                    }
                };

                var cEchoRequest = new DicomCEchoRequest();

                openAssociationRequest.AssociationToOpen.PresentationContexts.AddFromRequest(cEchoRequest);
                openAssociationRequest.AssociationToOpen.ExtendedNegotiations.AddFromRequest(cEchoRequest);

                DicomCEchoResponse cEchoResponse = null;
                await using (var association = await client.OpenAssociationAsync(openAssociationRequest, CancellationToken.None))
                {
                    await foreach (var response in association.SendRequestAsync(cEchoRequest, CancellationToken.None))
                    {
                        cEchoResponse = response;
                        break;
                    }
                }

                Assert.NotNull(cEchoResponse);
                Assert.Equal(DicomState.Success, cEchoResponse.Status.State);
            }
        }
```

Usage is quite verbose. This is because you get almost nothing for free. You are responsible for creating a DicomAssociation that supports C-ECHO requests. 

Opening an association is as simple as this:

```csharp
                await using (var association = await client.OpenAssociationAsync(openAssociationRequest, CancellationToken.None))
                {
                    ...
                }
```

When the association is asynchronously disposed (and it only implements IAsyncDisposable), it sends an association release request automatically, and waits for that to be confirmed by the server, with a timeout. 

Sending a request looks like this:

```csharp
                    await foreach (var response in association.SendRequestAsync(request, CancellationToken.None))
                    {
                        // Handle response. You can have multiple [Pending] responses, and then one [Completed] response
                    }
```

Sending a request returns `IAsyncEnumerable`, so you can easily loop through the responses as they come in. This actually makes the OnResponse and OnTimeout properties obsolete, and aligns the API more with HttpClient for example.

gofal commented:
Great the interface gives all the fexibility a user could need. Now this is also the perfect solution for my issue #652 :-)

gofal commented:
I guess it would need quite some amount of AssociationRequest-constructors that fill the properties with default- values. Like using the default DicomServerOptions if no explicit Options are set, or using the default NetworkConnection (registered via DI) of no explicit NetworkStreamConnection Options are set etc.
Otherwise it would be a lot of boilerplate code if you want to use the advanced client because of a single option you want to set explicit.

gofal commented:
Will this advanced DicomClient then exist beside the other DicomClient? I mean, will it be possible, to change the DicomClient to be only a Wrapper around the advanced dicomclient, so that there will only be one implementation to be maintained?

gofal commented on FO-DICOM.Core/Network/Client/Advanced/AdvancedDicomClient.cs:
```diff
@@ -0,0 +1,84 @@
+                            disposeConnection = false;
+                            return new AdvancedDicomClientAssociation(_advancedDicomClientOptions, connection, dicomAssociationAcceptedEvent.Association);
+                        case DicomAssociationRejectedEvent dicomAssociationRejectedEvent:
+                            throw new DicomAssociationRejectedException(dicomAssociationRejectedEvent.Result, dicomAssociationRejectedEvent.Source,
```
Maybe this should not be an exception. Because exceptions are some kind of unexpected errors. But in this case the advanced client receives a very valid message from the server saying that the association could not be estabished because of the reason XY. 
Do you think it would be tricky, to return here a DicomClientAssociation instance holding the information about the rejected request and having a property `Connected` that is false here?

the api would then be

await using (var association = await client.OpenAssociationAsync(openAssociationRequest, CancellationToken.None))
        {
            if (association.Connected)
            {
            await foreach (var response in association.SendRequestAsync(cEchoRequest, CancellationToken.None))
            {
                cEchoResponse = response;
                break;
            }
            } else { /* handle the rejected-case */ }
        }

gofal replied:
the SimpleDicomClient throws an exception, because the method `DicomClient.SendAsync()` has to handle everything and throws if something goes fail. there are no return values where the various states, that the client could be in, can be returned simply. But in this advanced client there you have several instances for the association, for the requests, etc. Here now there is the possiblility to return the states easily without throwing exceptions.

amoerie replied:
I have some doubts about this one. In my experience, association rejection is very rare, and almost always related to configuration issues (AE title not recognized, codecs not working, ...)

So I wonder if it's a good idea to change an API and force everyone to check a property on the association, while in 99% of the cases this property will be true. Furthermore, I think the API should be a "pit of success" where you are pushed towards correct usage. I think it would be too easy to forget to check this property and just assume the association is ready.

In my humble opinion, the exception makes sense: it _is_ exceptional that associations are rejected.

gofal replied:
you are using HttpClient API often as reference to go toward to. There an a request you get a response-object. this response-object has a http-status and the user has to check if response.IsInSuccessState. This HttpClient also throws only if there was no request-response for some reason (network,..), but if there was a response it does not throw - even if the response says 404 or similar.
Recently I noticed that the default DicomClient in version 5 has an event `AssociationRejected`, but that will never be raised, because the Send-method throws an exception which then aborts everything else.

amoerie replied:
I see where you are coming from, and your suggestion would definitely be feasible, but I do see a few subtle differences with HttpClient:

- An "accepted association" looks very different from a "rejected association". One has presentation contexts and a bunch of other metadata, the other only has a reject source and a reject reason. If we want to encode "association accepted" or "association rejected" as one single response type, we would have to find a good C# pattern to represent this. F# has the excellent "Either" type, which is implemented by some libraries in C# (see https://github.com/louthy/csharp-monad#either-monad), but otherwise you have to resort to an interface or abstract class and force the consumer to "type check", ie `if(associationResponse is AssociationAccepted) { ... } else if (associationResponse is AssociationRejected) { .. }`. I would prefer not to work with a single "Association.IsAccepted" property, because that object would have to contain all of the properties of association acceptance and rejection, and it would be confusing which property is filled in when.
 
- An association is necessary to _begin_ sending requests. You compare HTTP request/response with DICOM Association request/response, but I think you are comparing apples to oranges. HTTP request/response can already be compared with normal DICOM (ECHO / FIND / MOVE / ...) requests, where failure is indeed encoded in the object. (DicomStatus). I think a better comparison would be to compare the association handshake with the TLS handshake in HttpClient. If setting up the TLS connection to the server fails, HttpClient throws an exception. Because the TLS connection is a prerequisite to start sending requests. In fact, HttpClient negotiates the TLS algorithm with the server based on the capabilities of both the server and client. In a way, this can be compared to DICOM negotiating the presentation contexts.

amoerie replied:
> Recently I noticed that the default DicomClient in version 5 has an event AssociationRejected, but that will never be raised, because the Send-method throws an exception which then aborts everything else.

We could log a separate issue for this. But for me it also confirms that association rejection is quite rare.

gofal committed:
Merge branch 'development' into GH-1096-DicomClient

amoerie commented:
> I guess it would need quite some amount of AssociationRequest-constructors that fill the properties with default- values. Like using the default DicomServerOptions if no explicit Options are set, or using the default NetworkConnection (registered via DI) of no explicit NetworkStreamConnection Options are set etc.
> Otherwise it would be a lot of boilerplate code if you want to use the advanced client because of a single option you want to set explicit.

Better late than never, looking into this again.

I agree that we need to look at making this less boilerplate, although I also believe that the library shouldn't be too opinionated and provide too much here. It is, after all, the advanced DICOM client.

amoerie commented:
> Will this advanced DicomClient then exist beside the other DicomClient? I mean, will it be possible, to change the DicomClient to be only a Wrapper around the advanced dicomclient, so that there will only be one implementation to be maintained?

This is/was my intention, but I'm not sure whether I'll be able to do so without breaking changes. For example, the state pattern and GetNextState is one of the main reasons the call stacks end up being so deep, and I've completely avoided that problem now by publishing DICOM events into a channel. 

I would prefer to get rid of the internal states altogether maybe, but that would mean "OnStateChange" would likely be broken.

amoerie committed:
Try to build existing DicomClient on top of AdvancedDicomClient

amoerie committed:
Try and make some progress on the reuse of the advanced client from the regular DicomClient

amoerie committed:
Continue rebuilding DICOM client on top of advanced DICOM client

amoerie commented:
@gofal 

Update (yes I'm still working on this, although sporadic)

I'm continuing my work on rebuilding the existing DICOM client on top of the advanced DICOM client with as little API breakage as I possibly can. This means I have to reimplement a lot of existing features (association lingering, max requests per association).

The good news is that so far the API breakage is *very* limited. For example, the StateChanged event will still be completely supported. The only difference is that the state objects themselves will now be empty, since DicomClient no longer uses the state pattern for its implementation. 

I'm finally in a position where the tests are slowly but surely starting to turn green, but I think I have a lot of work ahead of me (manual testing, XML documentation, C-STORE and N-EVENT support, etc.)

This is what the unit tests look like at the moment:

![image](https://user-images.githubusercontent.com/2063584/119361266-c4ac7c80-bcab-11eb-90d6-4b60382550e7.png)

amoerie commented:
More discussion points:

- I've had to move away from implementing IAsyncDisposable on the association class, because IAsyncDisposable does not allow cancellation. And this is important, to implement proper timeout behavior on an association release / abort. Furthermore, it was hard to decide whether to abort or release the association in a disposal scenario. In the end, I felt it made more sense to simply expose a ReleaseAsync() and AbortAsync() method on the association, giving full control to the user

- For backwards compatibility reasons, I've had to split up the creation of the association in two calls. First, you have to call `advancedDicomClient.OpenConnectionAsync`. This gives you a connection. Then, you can use that connection to call `advancedDicomClient.OpenAssociationAsync`. The benefit here is that again, full control is given to the user to use separate cancellation tokens (i.e. separate timeouts), store the connection and association in different places, et cetera. The biggest pitfall I see is that this might give the impression that single connection can be used for multiple associations. This is false. In my experience, nearly every DICOM server actually closes the connection after an association release. 

I'm open to suggestions if we can make this API simpler again, although I have to admit I like the flexibility and the lack of abstraction

gofal commented:
Aren't those two calls `OpenConnectionAsync` and `OpenAssociationAsync` hidden from the user in default simple DicomClient class?
If a developer uses the advanced dicomclient then he has to care about himself. so if the then reuses the same connection for several associations and most servers close the connection after first association, then this developer will notice that and fix it? I doubt that a dev, who is not aware of the underlying connection and association architecture will try to use the advanced dicomclient. So for me that does not seem to be a pitfall.

amoerie commented:
> Aren't those two calls `OpenConnectionAsync` and `OpenAssociationAsync` hidden from the user in default simple DicomClient class?
> If a developer uses the advanced dicomclient then he has to care about himself. so if the then reuses the same connection for several associations and most servers close the connection after first association, then this developer will notice that and fix it? I doubt that a dev, who is not aware of the underlying connection and association architecture will try to use the advanced dicomclient. So for me that does not seem to be a pitfall.

Thank you, I share your opinion. When someone does try to do this, he will notice very quickly.
But I'm thinking if we can "tag" the connection as "already in use" so we can throw an exception when someone tries to reuse a connection. I think that wouldn't be too much effort, and would surface the problem a lot sooner with a clearer error message.

amoerie committed:
Ensure DicomClient always completes properly, even if the association is reused

Also extract the connection parameter from the OpenAssociationRequest, it makes more sense to have it standalone

amoerie commented:
Late night update 😁

![image](https://user-images.githubusercontent.com/2063584/119564358-fe16e200-bda8-11eb-8906-8669884bfd83.png)

amoerie committed:
Immediately stop lingering and begin sending when a new DICOM request arrives

amoerie committed:
Merge development into GH-1096-DicomClient

amoerie committed:
Improve DICOM network exceptions

Surface the root error when throwing exceptions after closing a connection

amoerie committed:
Avoid race condition in advanced association

Capture the cancellation token once in the ctor and keep reusing it
This avoids a subtle race issue where CollectEventAsync could throw if the cancellation token source is already disposed

amoerie committed:
Add support for C-GET and N-EVENTREPORT using advanced DICOM client

Provide a "RequestHandlers" property when opening the connection, and fill this in from the simple DicomClient

amoerie commented:
Thanks @mrbean-bremen for looking into it.

It looks like Github Actions heard me complain though, they did trigger now. While the tests are green on my machine, it looks like I still have some work to do..

amoerie committed:
Prevent all scenarios where DicomClient does not have a logger

amoerie committed:
Fix broken bug unit test

amoerie committed:
Improve error handling while waiting for responses

amoerie committed:
Improve logging

amoerie committed:
Nitpick: don't use string interpolation if not needed

amoerie committed:
Delete dummy file that turned out not to be needed

amoerie committed:
Delete the old state-based implementation of DicomClient

amoerie committed:
Guard against already disposed DicomService when trying to send or receive DICOM messages

amoerie committed:
Improve handling of already disposed DICOM association

amoerie committed:
Send DICOM requests in parallel if we can

amoerie committed:
Remove slightly too verbose logging

amoerie committed:
Handle request cancellation correctly

amoerie committed:
Expose IsDisposed on advanced DICOM client association so that we can avoid trying to abort it if the connection is already gone

amoerie committed:
Fix unit test where server is disposed: IO exceptions do not necessarily inherit from DicomNetworkException, and I am unsure if we always want to wrap these

amoerie commented:
All tests are green! That means DicomClient - which has been rewritten against the new advanced DICOM client - now fully supports all of the internal logic of the state based implementation. 🎉

amoerie committed:
Add documentation, copyright and other finishing touches

amoerie committed:
Add more docs and polish

amoerie committed:
Move the events into a single file

amoerie committed:
Add extensions for C-FIND, C-STORE and C-MOVE

amoerie committed:
Add test for C-FIND using advanced DICOM client

gofal commented:
@amoerie Thanks for this work. It is a huge piece of code. I will review it, but while I can review some small changes quite quickly, but to review this I will have to schedule some time.

amoerie committed:
Mini optimization: use singletons for events without parameters

amoerie committed:
Add more documentation and mark state-related APIs as obsolete

amoerie committed:
Remove unused class

amoerie committed:
Format code

amoerie committed:
Simplify and add missing docs in connection factory

amoerie committed:
Update changelog

amoerie commented:
I've (optimistically) added the advanced DICOM client to the changelog under 5.0.1, but I think that such a big change should be at least covered by a minor version bump (e.g. 5.1.0). I still need to check for breaking changes too.

amoerie committed:
Cleanup advanced DICOM client test a bit

amoerie committed:
Fix host in unit tests

amoerie committed:
Attempt to fix GPG key, please ignore

amoerie committed:
Attempt #2 to fix GPG key, please ignore

amoerie committed:
Merge branch 'GH-1096-DicomClient' of github.com:fo-dicom/fo-dicom into GH-1096-DicomClient

amoerie committed:
Add more XML documentation related to the new advanced DICOM client

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1096-DicomClient

amoerie committed:
Add sample usage of advanced DICOM client to README

amoerie committed:
Add XML docs to NetworkStreamCreationOptions

amoerie committed:
Apply some minor fixes to advanced association request

amoerie committed:
Make logger and options optional in advanced connection request

amoerie committed:
Fix usage of AdvancedDicomClientFactory

amoerie committed:
Add test coverage for open assocation and open connection

amoerie committed:
Add safety net so that a connection cannot be reused

This would fail later anyway, but undoubtedly with a more vague exception message

amoerie committed:
Add more missing XML docs

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1096-DicomClient

amoerie committed:
Improve disposal of DicomService and AdvancedDicomClientAssociation

amoerie committed:
Add test for C-STORE using advanced DICOM client

amoerie committed:
Add unit tests for C-MOVE

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1096-DicomClient

amoerie committed:
Fix disposal check in DicomService

amoerie committed:
Fix broken C-MOVE test

amoerie committed:
Fix race issue in DicomClient timeout test

amoerie committed:
Add C-MOVE test for advanced DICOM client

amoerie committed:
Remove debugging comments

amoerie committed:
Re-skip time based DicomClient tests to avoid reintroducing flakiness

amoerie commented:
Update: 

- No (major) breaking DicomClient API changes other than the fact that the OnStateChanged event now emits empty objects. The types still exist, but their properties are gone. Not sure if anyone was using this, the test definitely didn't.
- Added tests for C-STORE and C-MOVE
- Added some improvements for when the association or the connection gets disposed (because of a connection closure)

amoerie commented:
@gofal @mrbean-bremen this is now ready for review

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1096-DicomClient

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1096-DicomClient

amoerie committed:
Prevent events causing exceptions when connection is already closed

amoerie committed:
Reduce flakiness of DicomClient test

amoerie committed:
Add optimization for sudden disconnection in advanced associations

amoerie committed:
Immediately shutdown request channel when request completes

amoerie committed:
Don't keep rethrowing the same exception

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1096-DicomClient

amoerie commented:
I'll look into the failing test, don't merge yet.

amoerie committed:
Improve test annotations

amoerie committed:
Add escape hatch for unexpected disconnection/disposal

amoerie committed:
Merge development into
GH-1096-DicomClient

amoerie committed:
Do not allocate a TimeSpan all the time

amoerie committed:
Throw ObjectDisposedException when disposing an association with pending requests

amoerie committed:
Merge development into GH-1096-DicomClient

amoerie committed:
Add memory provider

amoerie committed:
Make separate DicomServiceDependencies and DicomServerDependencies classes

These will contain the services required by DicomServer and DicomService respectively. This will allow fo-dicom to change the dependencies in the future without impacting consumer code that inherits from these base classes. Unfortunately, this change does once require an intervention. The current passing through of ILogManager, INetworkManager and ITranscoderManager will have to be swapped with DicomServiceDependencies

Similarly, the ILogManager and INetworkManager dependencies of DicomServer derived classes will now have to forward an instance of DicomServerDependencies

amoerie committed:
Merge development into GH-1096-DicomClient

amoerie committed:
Implement association request timeout

amoerie commented:
Reopened to trigger CI :-)

amoerie committed:
Increase logging in ListenAndProcessPDUAsync

amoerie commented:
A failing (flaky) test in .NET Core 2.1 caused a CI failure. Upon deeper investigation, I think this looks a lot like #1325 where a badly timed SocketException causes the DicomClient to hang. 

I've been digging into the code trying to find out where it gets stuck. I'm fairly sure "TryCloseConnectionAsync" is a possible culprit, so I've added more debug level logging in places where this method is called.
This should allow us to gather more information for when this inevitably happens again.

amoerie committed:
Move advanced DICOM client changelog to the correct version

amoerie committed:
Avoid finalizer in DicomService if we can

amoerie committed:
Move OpenAssociationAsync from AdvancedDicomClient to AdvancedDicomClientConnection

At this point, it becomes clear that AdvancedDicomClient has no added value whatsoever, and can simply be removed.

amoerie committed:
Update docs and tests

amoerie commented:
After going through my changes one more time, I realized that the class "AdvancedDicomClient" was conceptually useless and that all of the useful work was being performed in AdvancedDicomClientConnection and AdvancedDicomClientAssociation.

Therefore, I decided to remove this meaningless in-between layer. After all is said and done, this is what the advanced API looks like today:

```csharp
var cancellationToken = CancellationToken.None;
// Alternatively, inject IDicomServerFactory via dependency injection instead of using this static method
using var server = DicomServerFactory.Create<DicomCEchoProvider>(12345); 

var connectionRequest = new AdvancedDicomClientConnectionRequest
{
    NetworkStreamCreationOptions = new NetworkStreamCreationOptions
    {
        Host = "127.0.0.1",
        Port = server.Port,
    }
};

// Alternatively, inject IAdvancedDicomClientConnectionFactory via dependency injection instead of using this static method
using var connection = await AdvancedDicomClientConnectionFactory.OpenConnectionAsync(connectionRequest, cancellationToken);

var associationRequest = new AdvancedDicomClientAssociationRequest
{
    CallingAE = "EchoSCU",
    CalledAE = "EchoSCP"
};

var cEchoRequest = new DicomCEchoRequest();

using var association = await connection.OpenAssociationAsync(associationRequest, cancellationToken);
try
{
    DicomCEchoResponse cEchoResponse = await association.SendCEchoRequestAsync(cEchoRequest, cancellationToken).ConfigureAwait(false);
    
    Console.WriteLine(cEchoResponse.Status);
}
finally
{
    await association.ReleaseAsync(cancellationToken);
}
```

In a nutshell, this is the flow:
1. Open a connection using `IAdvancedDicomClientConnectionFactory`
2. Open an association using the connection from step 1
3. Send one or more requests using the association from step 2

There are many more options on both AdvancedDicomClientConnectionRequest and AdvancedDicomClientAssociationRequest that are not shown here, but all are documented in C# and have sensible defaults where possible.
You can only open one association per connection, and this is automatically enforced.
`IAdvancedDicomClientAssociation` has extension method shorthands for the following DICOM requests: C-ECHO, C-FIND, C-MOVE and C-STORE.
Other types of requests are also supported of course, but you will need to use the generic `SendRequestAsync` which returns an  `IAsyncEnumerable<DicomResponse>`

amoerie committed:
Unblock DicomService when thread A is waiting to send PDUs while thread B processes a disconnection

amoerie committed:
Force close the connection when we are certain the network stream is unusable

amoerie committed:
Remove wrong documentation

amoerie committed:
Merge GH-1325-BlockedDicomService into GH-1096-DicomClient

amoerie committed:
Merge branch 'GH-1325-BlockedDicomService' into GH-1096-DicomClient

amoerie committed:
Merge GH-1325-BlockedDicomService into GH-1096-DicomClient

amoerie committed:
Merge GH-1325-BlockedDicomService into GH-1096-DicomClient

amoerie committed:
Add missing null check for event handler in DicomClient

amoerie commented:
Just a note here: this PR is ready for review. 

It includes the changes from #1332 (so it might make sense to merge that one first) because it solved the flaky test. 🎉

mrbean-bremen reviewed (commented):
Just to give some intermediate feedback here: I finally started to review this, and so far it all looks very neat. It may partly be a bit over my head (have to re-read some stuff in the standard, too), so I probably won't find problems with the code, but I will see to get further with the review within the next days anyway.

mrbean-bremen commented on FO-DICOM.Core/Network/Client/Advanced/Connection/AdvancedDicomClientConnectionEvents.cs:
```diff
@@ -0,0 +1,258 @@
+        /// <returns>
+        /// The <see cref="DicomCStoreResponse"/> related to the C-STORE <paramref name="request"/>.
+        /// </returns>
+        Task<DicomResponse> OnCStoreRequestAsync(DicomCStoreRequest request);
```
This (and the next one) look a bit out of place here, as the rest of the events is purely on association level, while these are about specific services, but I just may not get it.

mrbean-bremen replied:
Ok, this matches `AdvancedDicomClientConnectionRequestHandlers`, these are just the ones that are handled. I guess this is ok.

amoerie replied:
Yeah I had to keep this to support C-GET requests (where the server responds with a C-STORE request)
It's a bit out of place, and why most implementations use C-MOVE, but for backwards compatibility reasons we still need this.

mrbean-bremen replied:
Ah, right - I never had to do with C-GET and wasn't aware of this.

mrbean-bremen commented on FO-DICOM.Core/Network/Client/Advanced/Association/AdvancedDicomClientAssociation.cs:
```diff
@@ -0,0 +1,504 @@
+
+                            if (!requestChannel.Writer.TryWrite(requestCompletedEvent))
+                            {
+                                throw new DicomNetworkException(_responseChannelDoesNotHaveUnlimitedCapacity);
```
Can these really happen? So far, I see only unbounded channels created. Or can this be changed by the user?

amoerie replied:
No they should never happen. I wasn't too sure what to do here, but the idea is that if some future refactoring/change would change it to bounded channels, hopefully this exception message would show up ASAP.

mrbean-bremen commented on FO-DICOM.Core/Network/Client/Advanced/Association/AdvancedDicomClientAssociation.cs:
```diff
@@ -0,0 +1,504 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void ThrowIfAlreadyDisposed()
+        {
+            if (!IsDisposed)
```
Nitpicking coding style: everywhere else you use curly braces for statements in if clauses

amoerie replied:
You're right! Done.

mrbean-bremen commented on FO-DICOM.Core/Network/Client/Advanced/Association/AdvancedDicomClientAssociationRequest.cs:
```diff
@@ -0,0 +1,47 @@
+        /// <summary>
+        /// Gets the called application entity.
+        /// </summary>
+        public string CalledAE { get; set; }
```
Wouldn't this better be `CalledAET` or `CalledAETitle`? Same with `CallingAE`.
Though I now see that it's the same in `DicomAssociation`, so it is probably better to use the same names...

amoerie replied:
Yeah I wanted parity with DicomAssociation because that's what being made *from* this request.

mrbean-bremen commented on FO-DICOM.Core/Network/Client/Advanced/Connection/IAdvancedDicomClientConnectionInterceptor.cs:
```diff
@@ -0,0 +1,101 @@
+
+namespace FellowOakDicom.Network.Client.Advanced.Connection
+{
+    public interface IAdvancedDicomClientConnectionInterceptor
```
Maybe it would help to add a bit of description here.
Also, would it make sense to provide some default implementation that just calls the inner connection to avoid to have to implement each method?

mrbean-bremen replied:
On a second thought, the interceptor is instantiated internally, so this will probably not be feasible.

gofal replied:
agree, please more information here.
As I understand, the interceptor is forwarding all calls of incomming messages, but is has no default behavior. so what is the difference of setting an interceptor compared to register your own implementation of IAdvancedDicomClientConnection in DI?

amoerie replied:
As discussed in another comment as well, the interceptor implementation has been removed to trim down this merge request and allow for a more in-depth discussion for a possible future "interceptor" API in the future.

mrbean-bremen commented on FO-DICOM.Core/Network/Client/Advanced/Connection/DefaultAdvancedDicomClientConnectionFactory.cs:
```diff
@@ -0,0 +1,123 @@
+            if (request.Logger == null)
+            {
+                request.Logger = _logManager.GetLogger("Dicom.Network");
+            }
```
`request.Logger ??= _logManager.GetLogger("Dicom.Network");`

amoerie replied:
Done

mrbean-bremen commented on FO-DICOM.Core/Network/Client/Advanced/Connection/DefaultAdvancedDicomClientConnectionFactory.cs:
```diff
@@ -0,0 +1,123 @@
+            if (request.DicomServiceOptions == null)
+            {
+                request.DicomServiceOptions = _defaultDicomServiceOptions.Value;
+            }
```
same here

amoerie replied:
Done

mrbean-bremen commented on FO-DICOM.Core/Network/Client/DicomClient.cs:
```diff
@@ -1,26 +1,26 @@
 namespace FellowOakDicom.Network.Client
 {
-
     public interface IDicomClient
```
So this (e.g. `DicomClient` / `DefaultDicomClientFactory`) is a breaking change that demands a change of customer implementation, right?

amoerie replied:
Actually no! DicomClientFactory and DicomClient still work as they normally would and all the APIs should work.
The refactoring is entirely internal, since DicomClient is now built upon AdvancedDicomClientConnection / AdvancedDicomClientAssociation. 

If you look at the IDicomClient interface diff, I did remove the 4 "Manager" properties (LogManager / NetworkManager / ..) from IDicomClient since these have moved deeper into the stack and did not make a lot of sense being there. That's about the only "breaking" change I can think of.

mrbean-bremen replied:
I had hoped for that answer 😀 It looked to me that the ctor of the client and the factory method had changed, but I may have misread that. Will have another look a bit later...

mrbean-bremen reviewed (commented):
I'm stilll not finished (didn't get to it during the week), but it still looks quite good. Anyway, @gofal will have a deeper look... I will see if I can move on later today.

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1096-DicomClient

amoerie committed:
Fix code style in AdvancedDicomClientAssociation

amoerie committed:
Use ??= in DefaultAdvancedDicomClientConnectionFactory

mrbean-bremen commented on README.md:
```diff
@@ -48,7 +48,7 @@ Documentation, including API documentation, is available via GitHub pages:
 
 #### Image rendering configuration
-Out-of-the-box, *fo-dicom* for defaults to a internal class *FellowOakDicom.Imaging.IImage*-style image rendering. To switch to Desktop-style or ImageSharp-style image rendering, you first have to add the nuget packe you desire and then call:
+Out-of-the-box, *fo-dicom* for defaults to a internal class *FellowOakDicom.Imaging.IImage*-style image rendering. To switch to Desktop-style or ImageSharp-style image rendering, you first have to add the nuget package you desire and then call:
```
*fo-dicom* for defaults to -> *fo-dicom* defaults to?

amoerie replied:
Done!

mrbean-bremen reviewed (approved):
To me, this looks all good, but as I wrote, I'm out of my comfort zone here, so this is not a thorough review. Anyway, this is an impressive piece of work and looks very neat and consistent!

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1096-DicomClient

amoerie committed:
Update global.json to use .NET 6 SDK

amoerie committed:
Fix phrasing in README

amoerie committed:
Reinstate all benchmarks

amoerie committed:
Use latest version of System.Threading.Channels

amoerie committed:
Use stronger wording

amoerie committed:
Fix copyright headers in new files

amoerie committed:
Merge origin/development into GH-1096-DicomClient

amoerie committed:
Merge branch 'GH-1096-DicomClient' of github.com:fo-dicom/fo-dicom into GH-1096-DicomClient

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into GH-1096-DicomClient

amoerie committed:
Fix edge case

Fix edge case where a connection closed event still slipped through after the association was already disposed, resulting in an unobserved task exception in the association event collector

amoerie committed:
Ahum, actually run the benchmarks

amoerie committed:
Merge branch 'development' into GH-1096-DicomClient

amoerie committed:
Merge

amoerie committed:
Fix compilation issues after merge

amoerie commented:
@gofal Would you have some time for a last review?

gofal commented on FO-DICOM.Core/Network/Client/Advanced/Connection/AdvancedDicomClientConnection.cs:
```diff
@@ -0,0 +1,244 @@
+        }
+    }
+
+    public class InterceptingAdvancedDicomClientConnection : IAdvancedDicomClientConnection
```
please also add some description/information here, what the purpose of this interceptingAdvancedDicomClientconnection is and how/when to use this

amoerie replied:
After some more extensive testing with the interceptor, there were still some issues with it. For simplicity's sake, I've decided to remove this interceptor for now, because I believe its design should be discussed and reviewed separately. It's an important piece of the puzzle if you, for example, want to create a DICOM proxy, and my "simple" interceptor implementation here would not be a good tool for that.

Is it okay to review & merge the advanced DICOM client API separately? We can then still talk about and add a interceptor design afterwards. (If there is  even a need for this, I just assumed..)

gofal replied:
sure, lets do it that way.

gofal commented on FO-DICOM.Core/Network/Client/Advanced/Connection/AdvancedDicomClientConnection.cs:
```diff
@@ -0,0 +1,244 @@
+            Listener = Task.Run(RunAsync);
+        }
+
+        public new Task SendAssociationRequestAsync(DicomAssociation association) => base.SendAssociationRequestAsync(association);
```
What is the reason of overriding the methods and then only calling the base methods. Cannot those lines be skipped?

amoerie replied:
This is a technique to change the visibility of an inherited method. With "new", you can make a method public that is protected on the base class. When using this method, it does require your instance to be of type "AdvancedDicomClientConnection" (and not IDicomClientConnection)

These methods need to be public to satisfy the IDicomClientConnection interface.
Some methods like SendRequestAsync are already async, and in these cases it could be skipped yes, but I preferred the consistency of overriding every method then. This also ensures that any future change in method visibility would not have an impact here.

gofal commented on FO-DICOM.Core/Network/Client/Advanced/Connection/AdvancedDicomClientConnectionEvent.cs:
```diff
@@ -0,0 +1,234 @@
+    /// <summary>
+    /// When the TCP connection with the SCP is closed
+    /// </summary>
+    public class ConnectionClosedConnectionEvent : IAdvancedDicomClientConnectionEvent
```
nitpicking: having "connection" in the name twice is little bit too much, since this classes also live in the "connection" namespace. Wouldn't be "FellowOakDicom.Network.Client.Advanced.Connection.ConnectionClosedEvent" be sufficient? same with "DicomAbortedEvent" and "AssociationAcceptedEvent" and "RequestCompletedEvent" etc

amoerie replied:
You are completely right. In fact, this might be an artifact of a rename-refactoring I did. Nice catch! I'll change the names.

gofal commented on FO-DICOM.Core/Network/Client/Advanced/Connection/AdvancedDicomClientConnectionEvent.cs:
```diff
@@ -0,0 +1,234 @@
+            
+        }
+
+        private ConnectionClosedConnectionEvent(Exception exception)
```
why are those constructors private?

amoerie replied:
I reviewed these, and made a slight modification:

- Events with parameters (such as association rejected) have internal constructors. Only our code should produce these events.
- Events without parameters (such as association released) have internal singletons with a private constructor. 

These events are "internal plumbing" of the advanced DICOM client API and should not be used by consumers.
In the future, if we add some sort of interceptor API (see my comments above) we might need to expose these, but I'd prefer to keep them internal for now.

gofal replied:
ok, I see. I also thought that they should be public, if someone would register his own implementation of IAdvancedDicomClientConnection and therefore has to fire those events. But currently that is not designed.

gofal commented on FO-DICOM.Core/Network/Client/Advanced/Connection/AdvancedDicomClientConnectionEvents.cs:
```diff
@@ -0,0 +1,258 @@
+
+namespace FellowOakDicom.Network.Client.Advanced.Connection
+{
+    internal interface IAdvancedDicomClientConnectionEvents
```
is there a better name, that describes the responsibility of that interface/class?
There is the interface AdvancedDicomClientConnectionEvent, and then when I saw the AdvancedDicomClientConnectionEvents (just the plural), then i expected it to be something like ICollection<AdvancedDicomClientConnectionEvent>. but instead this class seems to be a processor/distributor/observer/whatever

amoerie replied:
The essence of what this class does is translating the callback based API (OnAssociationAccepted, OnAssociationReleased) into a collection of events. More specifically, it pushes them into a Channel<IAdvancedDicomClientConnectionEvent>, which is a sort of thread safe queue.
So I've renamed this to IAdvancedDicomClientConnectionEventCollector, if that's alright with you?

gofal replied:
thanks

gofal commented on FO-DICOM.Core/Network/Client/Advanced/Connection/AdvancedDicomClientConnectionRequest.cs:
```diff
@@ -0,0 +1,48 @@
+        /// The request handlers that will handle reverse requests.
+        /// For example, issuing a C-GET request will cause the other AE to send a C-STORE request back over the same association.
+        /// The C-STORE request handler configured here will be responsible for producing a C-STORE response 
+        /// Somewhat confusingly, C-GET temporarily swaps the roles of "DicomClient" and "DicomServer". Consider using C-MOVE requests instead.
```
c-move on the other hand causes lot of other issues: the server needs to have a lot of configuration in advance (mapping of AE-titles to ip/port), the server has to be able to open an association to the client which requires an proper network-konfiguration. the ip of the client therefore has to be fix (no dhcp), your application has to run an storescp in parallel to your dicomclient, .... 
I would not write such an one-sided recommendation into a documentation, that should just simply explain the behavior or the usage of a method or class.

mrbean-bremen replied:
See also [this blog entry](http://dclunie.blogspot.com/2016/05/to-c-move-is-human-to-c-get-divine.html) by @dclunie.

gofal replied:
great blog entry, thanks. The c-get is also the more intuitive to users. i constantly see questions in forums like: i did an c-move request, but the response I receive does not contain the requested dicom-file, there must be a bug in the library. :-)

amoerie replied:
I've removed the last line of my comment, sorry about that!

gofal commented on FO-DICOM.Core/Network/DesktopNetworkStream.cs:
```diff
@@ -35,32 +35,33 @@ public sealed class DesktopNetworkStream : INetworkStream
-            _tcpClient = new TcpClient { NoDelay = noDelay };
-            _tcpClient.ConnectAsync(host, port).Wait();
+            _tcpClient = new TcpClient { NoDelay = options.NoDelay };
+            _tcpClient.ConnectAsync(options.Host, options.Port).Wait();
```
i know, this is not related to the advanced client, but it came into mind: the options contain a timeout. shouldn't this timeout also be applied here con ConnectAsync?

amoerie replied:
See also my other comment:

> I'd be careful about that. Perhaps it should be a new property on NetworkStreamCreationOptions?
> 
> There's a conceptual difference between setting up a new connection (which might take a while) and sending a packet over an existing connection. I have seen site to site VPN connections take >5s for the first connection of the day, because there was some automatic spin up of VMs upon the first connection request.

I would prefer a new property specifically for the "initial connection timeout".
In fact, I also think using the AssociationRequestTimeoutInMs as a socket timeout is already a bit "wrong"...

gofal commented on FO-DICOM.Core/Network/Client/DicomClient.cs:
```diff
@@ -242,160 +227,433 @@ public class DicomClient : IDicomClient
+                                UseTls = UseTls,
+                                NoDelay = ServiceOptions.TcpNoDelay,
+                                IgnoreSslPolicyErrors = ServiceOptions.IgnoreSslPolicyErrors,
+                                Timeout = TimeSpan.FromMilliseconds(ClientOptions.AssociationRequestTimeoutInMs)
```
this AssociationRequestTimeout is passed to the networkstreamcreationoptions. but in DesktopNetworkStream this timeout is not applied to the tcpclient.connect method. maybe this timeout should be put into a cancellationtoken and then passed to OpenConnectionAsync?

amoerie replied:
I'd be careful about that. Perhaps it should be a new property on NetworkStreamCreationOptions? 

There's a conceptual difference between setting up a new connection (which might take a while) and sending a packet over an existing connection. I have seen site to site VPN connections take >5s for the first connection of the day, because there was some automatic spin up of VMs upon the first connection request.

gofal commented on ChangeLog5.md:
```diff
@@ -9,6 +9,8 @@
 * Fixed bug where disposal of DicomService could throw an exception if the buffered write stream still had content (#1319)
 * Improve handling of dropped connections (#1332)
+* Add new API for expert DICOM client connections/associations. This new API gives full control over the connection, association and requests when sending DICOM requests (#1144)
+* Rewrite "DicomClient" based on the new advanced API. The state based implementation is gone. The public API has remained the same. (#1144)
```
please also add some note about the (minimal) breaking change: that the dicomclient does not have the various managers as parameter and propery any more.

amoerie replied:
Done!

gofal commented on FO-DICOM.Core/Setup.cs:
```diff
@@ -91,6 +93,7 @@ public static IServiceCollection AddDicomClient(this IServiceCollection services
         {
             services.TryAddSingleton<DicomServiceDependencies>();
             services.TryAddSingleton<IDicomClientFactory, DefaultDicomClientFactory>();
+            services.TryAddSingleton<IAdvancedDicomClientConnectionFactory, DefaultAdvancedDicomClientConnectionFactory>();
```
I noticed, you created interfaces for all advanced classes that the dicomclient now uses, so I expected you did that to make use of DI.
Is there a reason, why you then create all those classes directly instead registering them here and make use of DI?

amoerie replied:
I'll walk through them:

- `IAdvancedDicomClientConnectionFactory`: this is a public API to create instances of `IAdvancedDicomClientConnection`. For public APIs, I prefer to use interfaces for maximum future flexibility. This factory can be initialized from DI, and does so.
- `IAdvancedDicomClientConnection`: this is a public API that advanced consumers can obtain via `IAdvancedDicomClientConnectionFactory.OpenConnectionAsync`. A connection cannot be registered via DI because the NetworkStream instance is dynamic.
- `IAdvancedDicomClientAssociation`: this is a public API that you get from `IAdvancedDicomClientConnection.OpenAssociationAsync`. An association cannot be registered via DI because the connection is dynamic.
-  `IAdvancedDicomClientConnectionEventCollector`: this is an internal API where we could maybe remove the interface and just use the class instead. But it cannot be registered in DI because the request handlers are dynamic (OnCStoreRequest and OnNEventRequest)

The only opportunity I see is IAdvancedDicomClientConnectionEventCollector, which could be "simplified". Even though the interface is only "for us", I do like the fact that the interface contains all of the signatures with their XML documentation in one place. It gives a "helicopter view" of what the event collector does.

gofal reviewed (commented):
see my comments above. Thanks!

amoerie committed:
Merge branch 'GH-1096-DicomClient' of github.com:fo-dicom/fo-dicom into GH-1096-DicomClient

amoerie committed:
Remove interceptor in advanced DICOM client

It's not fully working anyway, and the use case is highly unclear

We should look at what problems can be solved with such an interceptor, and if an interceptor is still the best design to solve those problems.

amoerie committed:
Simplify internal advanced DICOM client event type names

amoerie committed:
Make the advanced DICOM events constructable only from internal code

amoerie committed:
Rename IAdvancedDicomClientConnectionEvents to IAdvancedDicomClientConnectionEventCollector

amoerie committed:
Remove opinionated comment against C-GET

amoerie committed:
Fix docs of various DicomClientOptions

amoerie committed:
Document breaking change in DicomClient

amoerie commented:
@gofal Thanks for the thorough review! I think I went through all your comments. Can you see if my answers/changes resolve them?
